### ISSUE TITLE
SerilogInput ToRawValue() doesn't recurse, so forces destructuring to only one level

### ISSUE THREAD

**Issue by BuddhaBuddy1** [Azure/diagnostics-eventflow#348]
Is there a reason SerilogInput.ToRawValue() ignores Serilog's depth setting? I don't really see the reason for ToScalarValue() to exist. I'm trying to destructure a field in a destructured object, and SerilogInput is turning it into a string.

**Comment by karolz-ms** [comment]
The main reason for this code was to avoid issues with object graphs that include circular dependencies. 

What are you referring to, specifically, when you mention "Serilog depth setting"? The current implementation was contributed by nblumhardt (Serilog author), but that was 3 years ago, so things might have changed since...

**Comment by BuddhaBuddy1** [comment]
Serilog is configurable as far as how deep it will go into destructuring, to avoid circular dependencies. The default used to be 10, but was reduced to 5 a little while ago, I think. I wasn't sure why my object wasn't being destructured, so I tried setting it manually, like this:
```
            Logger = new LoggerConfiguration()
                .WriteTo.EventFlow(loggingOptions.DiagnosticPipeline)
                .Destructure.ToMaximumDepth(5)
                .Destructure.With<JsonNetDestructuringPolicy>()
                .Enrich.FromLogContext()
                .CreateLogger()
```
But it wasn't working. Only my top level item was being destructured, even though I knew my destructuring code was being hit, until I saw this code.

**Comment by karolz-ms** [comment]
I understand the question but, honestly, I do not know enough about Serilog to have some ideas how to improve existing `SerilogInput` code to do a better job converting Serilog entities such as `StructureValue` into set of vanilla .NET dictionaries and arrays, types that the rest of EventFlow pipeline would understand.

@nblumhardt any recommendations? The relevant code is here https://github.com/Azure/diagnostics-eventflow/blob/master/src/Microsoft.Diagnostics.EventFlow.Inputs.Serilog/SerilogInput.cs

**Comment by nblumhardt** [comment]
Thanks for the ping @karolz-ms.

Unfortunately it was too long ago now for me to remember why the conversion isn't arbitrarily deep. It looks to me like recursing on `ToRawValue` rather than `ToRawScalar` should be fine, but I have no real familiarity with EventFlow. Could be a good opportunity to experiment, for someone :-)

Cheers!

**Comment by karolz-ms** [comment]
Totally understand, thanks for looking into this @nblumhardt !

@BuddhaBuddy1 if you want, we could collaborate on this a bit and see if we can come up with something. I suggest we start with a few examples of simple- and not-so-simple data that you want to be logged and how you think they should appear to EventFlow outputs after destructuring. In other words, how about a bit of TDD? ☺️

**Comment by BuddhaBuddy1** [comment]
I had been having problems debugging since I upgraded to Visual Studio 14.4, so I downgraded to 13.10, and now the refactoring to remove ```ToRawScalar``` and call ```ToRawValue``` recursively is working! I feel like something else must be the cause, but I don't know what. It doesn't look like I expected in Kibana, but the JSON tab looks correct. In the Table view, though, it is just showing the JSON for the entire object. However, I can filter off of the individual keys and values, which is what I wanted.

I added a new parameter to ```ToRawValue(LogEventPropertyValue logEventValue, int depth)```, and replace all existing calls to ```ToRawScalar``` with a call to ```ToRawValue(xxx, depth - 1)```, and I also added two lines at the very beginning:
```
            if (depth <= 1)
                return ToRawScalar(logEventValue);
```
The first call passes in a hard coded depth of 5, because I couldn't figure out how to access the depth from what was passed in, but honestly, I don't think the depth code is necessary, as it looks like Serilog takes care of that itself in PropertyValueConverter.GetProperties(). I think if you just remove ```toRawScalar()``` and replaced all calls to it with calls to ```ToRawValue``` would be good enough.

**Comment by BuddhaBuddy1** [comment]
I'm not sure how to do a pull request. I cloned it locally, but I don't have permission to commit my local branch to GitHub. I just removed ```ToRawScalar()``` and replaced with calls to ```toRawValue()```, and added a test to SerilogInputTest:
```
        [Fact]
        public void DestructureDepthGreaterThanOne()
        {
            var healthReporterMock = new Mock<IHealthReporter>();
            var observer = new Mock<IObserver<EventData>>();
            using (var serilogInput = new SerilogInput(healthReporterMock.Object))
            using (serilogInput.Subscribe(observer.Object))
            {
                var logger = new LoggerConfiguration().WriteTo.Sink(serilogInput).CreateLogger();

                var structure = new { A = "alpha", B = "bravo", C = new { Y = "yankee", Z = "zulu" } };
                logger.Information("Here is {@AStructure}", structure);

                observer.Verify(s => s.OnNext(It.Is<EventData>(data =>
                    ((IDictionary<string, object>)data.Payload["AStructure"])["A"].Equals("alpha") &&
                    ((IDictionary<string, object>)data.Payload["AStructure"])["B"].Equals("bravo") &&
                    ((IDictionary<string, object>)((IDictionary<string, object>)data.Payload["AStructure"])["C"])["Y"].Equals("yankee") &&
                    ((IDictionary<string, object>)((IDictionary<string, object>)data.Payload["AStructure"])["C"])["Z"].Equals("zulu")
                )));
            }
        }
```

**Comment by karolz-ms** [comment]
Sounds like a great start! :+1: 

To do a pull request on GitHub you fork the repo to your account (the button in the top-right corner) and then commit all the changes there. I will create a branch in the main repo for us to collaborate (e.g. `feature/serilog-structured-data`). Please submit your PR from your fork to that branch (not `master`).

**Comment by karolz-ms** [comment]
Branch created https://github.com/Azure/diagnostics-eventflow/tree/feature/serilog-structured-data

**Comment by BuddhaBuddy1** [comment]
Done. I used a configuration setting to change the behavior, so it continues to follow current behavior unless otherwise configured. I also change how my tests worked from how the current tests work, as the current test methods only show THAT something failed, not WHY it failed.

**Comment by karolz-ms** [comment]
Thank you!! I will look at your PR shortly, sorry for the delay, busy week starting 😬

**Comment by BuddhaBuddy1** [comment]
No problem! Sorry for the delay in signing the CLA. We don't have any official allowances for submitting code to open source projects, and the person I needed permission from was on vacation until today.

**Comment by BuddhaBuddy1** [comment]
Do you know how I would configure the new setting? Is it possible to put it in the Serilog section of the "inputs" section of eventFlowConfig.json, like this:
```
{
    "type": "Serilog",
    ""ignoreSerilogDepthLevel"": false
}
```
Or do I have to use ```ReadFrom.Configuration()``` with a brand new config file?

**Comment by BuddhaBuddy1** [comment]
I assumed that configurations from eventFlowConfig.json automatically were included in the config, and updated the documentation for the Serilog component. I'll update again if that is not the case. Feel free to update the text, as I am not much of a writer.

I really wasn't sure about reversing the meaning of the config setting and making the default true to keep backward compatibility. If some other config setting name fits better with the rest of the libraries, feel free to change that, as well.

**Comment by karolz-ms** [comment]
@BuddhaBuddy1 just FYI I am running some tests and I don't quite like the behavior when object trees are passed to the input. I mean things when an entity has a collection of children, and each child has a collection of grandchildren, and so on. I am not entirely sure yet what the correct output should be in this case, but it is probably not what I am seeing with the current implementation. So it will take a few days for me to figure it out.

Bottom line, I am working on it ☺️

**Comment by BuddhaBuddy1** [comment]
Sure, I should have tested it with more complex objects. I only went two layers deep. I'm relocating my home office and won't be able to take a look at it myself until Monday. However, Serilog itself isn't all that great at destructuring objects OOTB. I found that I generally had to add a custom destructuring helper for things that I thought it should have figured out. If you are trying a real world test, try adding the Destructirama JsonNet library. https://www.nuget.org/packages/Destructurama.JsonNet/

**Comment by BuddhaBuddy1** [comment]
It works as I would expect in a test class. Lists and arrays both get turned into arrays of object, and objects themselves get turned into IDictionary<string, object>, which allows for one of them being a List or array. Here are some tests, including one that tests Destructure.ToMaximumDepth():
```

        [Fact]
        public void DestructureNestedArrays()
        {
            var healthReporterMock = new Mock<IHealthReporter>();
            var observer = new Mock<IObserver<EventData>>();
            IDictionary<string, object> spiedPayload = null;
            observer.Setup(p => p.OnNext(It.IsAny<EventData>()))
                .Callback<EventData>((p) => { spiedPayload = p.Payload; });
            var inMemorySettings = new Dictionary<string, string>
            {
                {SerilogInput.IGNORE_SERILOG_DEPTH_LEVEL_CONFIG, "false"},
            };

            IConfiguration configuration = new ConfigurationBuilder()
                .AddInMemoryCollection(inMemorySettings)
                .Build();
            using (var serilogInput = new SerilogInput(healthReporterMock.Object, configuration))
            using (serilogInput.Subscribe(observer.Object))
            {
                var logger = new LoggerConfiguration().WriteTo.Sink(serilogInput).CreateLogger();

                var structure = new { A = "alpha", B = new[] { new {Y = "yankee" , Z = new[] { new { Alpha  = "A" } } }, new {Y = "zulu", Z = new[] { new { Alpha = "B" } } } } };
                logger.Information("Here is {@AStructure}", structure);
            }
            Assert.Equal("alpha", ((IDictionary<string, object>)spiedPayload["AStructure"])["A"]);
            object b = ((IDictionary<string, object>)spiedPayload["AStructure"])["B"];
            Assert.True(b is object[]);
            Assert.Equal(2, ((object[])b).Length);
            var bb = (IDictionary<string, object>)((object[])b)[0];
            Assert.Equal("yankee", bb["Y"]);
            object z = bb["Z"];
            Assert.True(z is object[]);
            Assert.Single((object[])z);
            Assert.True(((object[])z)[0] is IDictionary<string, object>);
            var zz = (IDictionary<string, object>)((object[])z)[0];
            Assert.Equal("A", zz["Alpha"]);
            Assert.True(((object[])b)[1] is IDictionary<string, object>);
            bb = (IDictionary<string, object>)((object[])b)[1];
            Assert.Equal("zulu", bb["Y"]);
            z = bb["Z"];
            Assert.True(z is object[]);
            Assert.Single((object[])z);
            Assert.True(((object[])z)[0] is IDictionary<string, object>);
            zz = (IDictionary<string, object>)((object[])z)[0];
            Assert.Equal("B", zz["Alpha"]);
        }

        [Fact]
        public void DestructureNestedArraysPastDepthLimit()
        {
            var healthReporterMock = new Mock<IHealthReporter>();
            var observer = new Mock<IObserver<EventData>>();
            IDictionary<string, object> spiedPayload = null;
            observer.Setup(p => p.OnNext(It.IsAny<EventData>()))
                .Callback<EventData>((p) => { spiedPayload = p.Payload; });
            var inMemorySettings = new Dictionary<string, string>
            {
                {SerilogInput.IGNORE_SERILOG_DEPTH_LEVEL_CONFIG, "false"},
            };

            IConfiguration configuration = new ConfigurationBuilder()
                .AddInMemoryCollection(inMemorySettings)
                .Build();
            using (var serilogInput = new SerilogInput(healthReporterMock.Object, configuration))
            using (serilogInput.Subscribe(observer.Object))
            {
                var logger = new LoggerConfiguration().WriteTo.Sink(serilogInput).Destructure.ToMaximumDepth(4).CreateLogger();

                var structure = new { A = "alpha", B = new[] { new { Y = "yankee", Z = new[] { new { Alpha = "A" } } }, new { Y = "zulu", Z = new[] { new { Alpha = "B" } } } } };
                logger.Information("Here is {@AStructure}", structure);
            }
            Assert.Equal("alpha", ((IDictionary<string, object>)spiedPayload["AStructure"])["A"]);
            object b = ((IDictionary<string, object>)spiedPayload["AStructure"])["B"];
            Assert.True(b is object[]);
            Assert.Equal(2, ((object[])b).Length);
            var bb = (IDictionary<string, object>)((object[])b)[0];
            Assert.Equal("yankee", bb["Y"]);
            object z = bb["Z"];
            Assert.True(z is object[]);
            Assert.Single((object[])z);
            Assert.Null(((object[])z)[0]);
            Assert.True(((object[])b)[1] is IDictionary<string, object>);
            bb = (IDictionary<string, object>)((object[])b)[1];
            Assert.Equal("zulu", bb["Y"]);
            z = bb["Z"];
            Assert.True(z is object[]);
            Assert.Single((object[])z);
            Assert.Null(((object[])z)[0]);
        }

        [Fact]
        public void DestructureNestedLists()
        {
            var healthReporterMock = new Mock<IHealthReporter>();
            var observer = new Mock<IObserver<EventData>>();
            IDictionary<string, object> spiedPayload = null;
            observer.Setup(p => p.OnNext(It.IsAny<EventData>()))
                .Callback<EventData>((p) => { spiedPayload = p.Payload; });
            var inMemorySettings = new Dictionary<string, string>
            {
                {SerilogInput.IGNORE_SERILOG_DEPTH_LEVEL_CONFIG, "false"},
            };

            IConfiguration configuration = new ConfigurationBuilder()
                .AddInMemoryCollection(inMemorySettings)
                .Build();
            using (var serilogInput = new SerilogInput(healthReporterMock.Object, configuration))
            using (serilogInput.Subscribe(observer.Object))
            {
                var logger = new LoggerConfiguration().WriteTo.Sink(serilogInput).CreateLogger();

                var structure = new { A = "alpha", B = Outer.Construct(new string[][] { new string[] { "yankee", "A" }, new string[] { "zulu", "B"} }) };
                logger.Information("Here is {@AStructure}", structure);
            }
            Assert.Equal("alpha", ((IDictionary<string, object>)spiedPayload["AStructure"])["A"]);
            object b = ((IDictionary<string, object>)spiedPayload["AStructure"])["B"];
            Assert.True(b is object[]);
            Assert.Equal(2, ((object[])b).Length);
            var bb = (IDictionary<string, object>)((object[])b)[0];
            Assert.Equal("yankee", bb["Y"]);
            object z = bb["Z"];
            Assert.True(z is object[]);
            Assert.Single((object[])z);
            Assert.True(((object[])z)[0] is IDictionary<string, object>);
            var zz = (IDictionary<string, object>)((object[])z)[0];
            Assert.Equal("A", zz["Alpha"]);
            Assert.True(((object[])b)[1] is IDictionary<string, object>);
            bb = (IDictionary<string, object>)((object[])b)[1];
            Assert.Equal("zulu", bb["Y"]);
            z = bb["Z"];
            Assert.True(z is object[]);
            Assert.Single((object[])z);
            Assert.True(((object[])z)[0] is IDictionary<string, object>);
            zz = (IDictionary<string, object>)((object[])z)[0];
            Assert.Equal("B", zz["Alpha"]);
        }

        class Outer
        {
            public Outer( string[] list)
            {
                Y = list[0];
                for (int i = 1; i < list.Length; i++)
                    Z.Add(new Inner() { Alpha = list[i] });
            }
            public string Y { get; set; }
            public List<Inner> Z { get; set; } = new List<Inner>();
            public static List<Outer> Construct(string[][] list)
            {
                List<Outer> retVal = new List<Outer>();
                foreach (var item in list)
                    retVal.Add(new Outer(item));
                return retVal;
            }
        }

        class Inner
        {
            public string Alpha { get; set; }
        }
```

**Comment by karolz-ms** [comment]
Just looking at the first test, I would expect `B` to be an array, not a dictionary...

**Comment by karolz-ms** [comment]
Looks like when writing my test, I have set the destructuring depth to shallow! :satisfied:

**Comment by karolz-ms** [comment]
I would argue that `DestructureNestedArraysPastDepthLimit` is an unexpected result. The input **should not** do this. I have an implementation in a private branch that works differently, and IMO, correctly

**Comment by karolz-ms** [comment]
All right @BuddhaBuddy1 , I have a "release candidate" ready. Please check out https://github.com/Azure/diagnostics-eventflow/commit/fb30b98357e0e2d2bc7cc28f2b19635f371de4e6 and let me know what you think. 

I did some refactoring but the core is still 100% your idea of using `ToRawValue` recursively, which worked out really well.

**Comment by BuddhaBuddy1** [comment]
I knew there had to be a better way to do the config. :-)

Regarding the test going beyond the depth limit, I'm not sure how it can not be the expected result, since Serilog did it. If we are changing what Serilog gives us, then I think we are breaking a contract with Serilog, and possibly making things confusing for experienced Serilog users.

You don't say what is unexpected, but from your candidate, it looks like the only difference is that you are removing the field completely when the value is null, whether it is because it went beyond the destructuring depth, or it is just a null value (as far as I know there is no way for us to tell the difference). This will make the output for current users different that what they get now, even if they don't use the new feature. If someone has a monitor counting how often a field has an empty value, once they update they won't have that field at all. If someone wants nulls removed, they can create a IDestructurePolicy to remove them. I don't think we should do it for them and not give a way to opt out.

If nothing else, the empty value is a clue that there WAS something in that field, but it wasn't logged because it went to deep into the object tree.

**Comment by karolz-ms** [comment]
Thanks for the feedback! There are two issues here 

*Eliminating null values from arrays* <br/> 
I think you are making a good argument for not doing that. The reasons why I initially decided to do so are as follows. First, I believe in most cases they will cause more trouble than benefit down the line (i.e. when the data hits the destination and is analyzed/reported on--I do not have a high confidence that they will be handled properly). Second, with the depth limiting you cannot really distinguish between the case when a property had no value in the first place vs. when a property is set to null by depth-limiting logic. So the utility of having a bunch of nulls in an array is low to begin with. That said, I think you are right that is better to maximize fidelity with Serilog, so I will revert that change.

*Destructuring logic going beyond depth limit* <br/>
If this is indeed happening, it is clearly wrong and going against the core idea of the destructuring depth limit. If Serilog does it, and not SerilogInput code, then it is a bug in Serilog. I will take a closer look.

**Comment by karolz-ms** [comment]
I think we are good to go https://github.com/Azure/diagnostics-eventflow/pull/351/commits/fcb352680877ee6cd1543b2697128e40ccfd7595 After refraining from removing null values from arrays the destructuring depth still works as expected ☺️

**Comment by BuddhaBuddy1** [comment]
Thanks a ton!

**Comment by karolz-ms** [comment]
Should be available on nuget.org now (ver 1.5.0)

Could you let me know if it works as expected? TIA!

**Comment by BuddhaBuddy1** [comment]
Yes, perfectly! Thank you!

**Comment by BuddhaBuddy1** [comment]
One caveat (that seems pretty standard for ELK) is that if your structure changes, you need to refresh your index pattern, or your messages are silently ignored. Something to keep in mind when someone changes their code to use the new setting.

**Comment by karolz-ms** [comment]
😌 good point. 

Thanks again for the idea, implementation, and for verifying that the new bits work!
